networking/notes/shelley/CommunicationPatterns.md

```diff
@@ -0,0 +1,283 @@
+  - Keepalive : send after a certain amount of inactivity. Useful for
+      detecting half-open TCP connections (intermittent lower-layer failures).
+
+### Publisher sends
+
+  - Announce key : indicate that a payload for this key is available.
+  - Payload value : give the payload for a key.
+
+Should the publisher acknowledge add/remove topic?
+If a subscriber adds a topic and then removes it, there may be announces for
+that topic inbound.
+
+## Interface
+
+```Haskell
+data PubSub (topic :: * -> *) d
+
+-- Every topic, with key and value types given.
+data Topic k v where
+  Block :: Topic BlockHeader Block
+  Txp :: Topic TxId TxAux
+  Mpc :: Topic StakeholderId MpcMsg
+  ...
+
+-- Subscriber can send values of this type.
+data Subscription topic where
+  AddTopic :: topic k v -> Subscription topic
+  RemoveTopic :: topic k v -> Subscription topic
+  Request :: topic k v -> k -> Subscription topic
+  Ignore :: topic k v -> k -> Subscription topic
+  Keepalive :: Subscription topic
+
+-- Publisher can send values of this type.
+data Publication topic where
+  Announce :: topic k v -> k -> Publication topic
+  Payload :: topic k v -> v -> Publication topic
+
+-- Publish a message. Asynchronous. Returns right away.
+-- The program which publishes is not informed of publication failure.
+-- The pub/sub system can retry or drop according to some policy.
+publish :: PubSub topic d -> Publication topic -> d ()
+
+-- Does nothing if already subscribed and returns False.
+-- The callback is a mechanism for backpressure. The dequeueing thread of the
+-- pub/sub system will wait for the callback to finish before continuing.
+subscribe :: PubSub topic d -> topic k v -> d Bool
+
+-- Does nothing if not currently subscribed.
+unsubscribe :: PubSub topic d -> topic k v -> d ()
+
+-- What to do with a publication.
+-- Announcements must indicate whether to request or ignore.
+-- Payloads must be processed somehow.
+data Next topic d = Next
+  { announce :: NodeId -> topic k v -> k -> d Bool
+  , payload :: NodeId -> topic k v -> v -> d ()
+  }
+
+-- Take the next publication (from any subscriber).
+next :: PubSub topic d -> Next topic d -> d ()
+```
+
+## Questions, caveats, risks
+
+### Publish over subscription channel
+
+It would make a lot of sense to send the published data over the same
+conversation which is used to establish the subscription, rather than (as we
+do now) have one conversation control the lifetime of the connection, and
+starting other conversations to publish each datum.
+
+It's not clear whether this has any big advantages or disadvantages.
+
+Decided: we'll use one time-warp conversation for the whole subscription.
+The subscriber can add/remove topics, ACK published keys, and send keepalive
+messages. The publisher can send keys and values.
+
+### Changes to time-warp
+
+The goal should be to do minimal changes to time-warp. It already enables
+what we need to do, but not in the most efficient or straightforward way.
```

Comment:
@dcoutts said

> There will be a new requirement (eventually, not necessarily top priority) to be able to limit the bandwidth that all other peers can force this node into providing. Consider running a node at home without a firewall so that other nodes can contact you. That's fine so long as it does not consume all your home DSL upload bandwidth, otherwise you'd be annoyed. So we would need to be able to cap that. Timewarp does know (in principle) the amount of data sent to peers as responses to conversations, so it could in principle impose an aggregate bandwidth cap.